SSA/Projects/InstCombine/HackersDelight.lean

```diff
@@ -459,6 +459,150 @@ theorem div_overflow_iff_neq_and_RightShift_lt {x y : BitVec 64} {y : BitVec 32}
+
+def unsignedMinBitVec (x y : BitVec w) : BitVec w :=
+  if x ≤ᵤ y then x else y
+
+def signedMaxBitVec (x y : BitVec w) : BitVec w :=
+  if x ≥ₛ y then x else y
+
+def signedMinBitVec (x y : BitVec w) : BitVec w :=
+  if x ≤ₛ y then x else y
+
+theorem signed_max_eq_add_doz :
+    signedMaxBitVec x y = y + signedDifferenceOrZero x y := by
+  try alive_auto
+  all_goals sorry
+
+theorem signed_min_eq_sub_doz :
+    signedMinBitVec x y = x - signedDifferenceOrZero x y := by
+  try alive_auto
+  all_goals sorry
+
+theorem unsigned_max_eq_add_dozu :
+    unsignedMaxBitVec x y = y + unsignedDifferenceOrZero x y := by
+  try alive_auto
+  all_goals sorry
+
+theorem unsigned_min_eq_sub_dozu :
+    unsignedMinBitVec x y = x - unsignedDifferenceOrZero x y := by
+  try alive_auto
+  all_goals sorry
+
+def leBitmask (x y : BitVec w) : BitVec w :=
+  if y ≤ x then -1#w else 0#w
+
+theorem signed_doz_eq_sub_and_bitmask :
+   signedDifferenceOrZero x y = (x - y) &&& leBitmask x y := by
+  try alive_auto
+  all_goals sorry
+
+theorem signed_max_xor_and_bitmask_xor :
+    signedMaxBitVec x y = ((x ^^^ y) &&& leBitmask x y) ^^^ y := by
+  try alive_auto
+  all_goals sorry
+
+theorem signed_min_xor_and_bitmask_xor :
+    signedMinBitVec x y = ((x ^^^ y) &&& leBitmask y x) ^^^ y := by
+  try alive_auto
+  all_goals sorry
+
+def carryBitmask (x y : BitVec w) : BitVec w :=
+  if BitVec.carry w x (~~~y) false then -1#w else 0#w
+
+theorem unsigned_doz_sub_and_not_carry :
+    unsignedDifferenceOrZero x y = ((x - y) &&& ~~~ carryBitmask x y) := by
+  try alive_auto
+  all_goals sorry
+
+theorem unsigned_max_eq_sub_sub_and_carry :
+    unsignedMaxBitVec x y = x - ((x - y) &&& carryBitmask x y) := by
+  try alive_auto
+  all_goals sorry
+
+theorem unsigned_min_eq_add_sub_and_carry :
+    unsignedMinBitVec x y = y + ((x - y) &&& carryBitmask x y) := by
+  try alive_auto
+  all_goals sorry
+
+section fixedWidth
+variable {x y d : BitVec 32}
+
+theorem signed_doz_and_not_xor_xor_and_and_rightShift (h : d = x - y) :
+    signedDifferenceOrZero x y = d &&& (~~~ d ^^^ ((x ^^^ y) &&& (d ^^^ x)) >>> 31) := by
+  try alive_auto
+  all_goals sorry
+
+theorem unsigned_doz_and_not_xor_xor_and_and_rightShift (h : d = x - y) :
+    unsignedDifferenceOrZero x y = d &&& ~~~ (((~~~ x &&& y) ||| ~~~ (x ^^^ y) &&& d) >>> 31) := by
+  try alive_auto
+  all_goals sorry
+
+theorem signed_doz_eq_sub_and_not_sub_rightShift :
+    signedDifferenceOrZero x y = (x - y) &&& ~~~ ((x - y) >>> 31) := by
+  try alive_auto
+  all_goals sorry
+
+theorem signed_max_eq_sub_sub_and_sub_rightShift :
+    signedMaxBitVec x y = x - ((x - y) &&& ((x - y) >>> 31)) := by
+  try alive_auto
+  all_goals sorry
+
+theorem signed_min_eq_add_sub_and_sub_rightShift :
+    signedMinBitVec x y = y + ((x - y) &&& ((x - y) >>> 31)) := by
+  try alive_auto
+  all_goals sorry
+
+theorem lt_sdoz_or_neg_sdoz :
+    (y <ₛ x) ↔ (signedDifferenceOrZero x y ||| -(signedDifferenceOrZero x y)).getMsb 32 := by
```

Comment:
I don't know what the statement of the theorem in Hacker's Delight is, but I think it is probable that you want  "BitVec.msb" rather than "BitVec.getMsb 32"

Reply:
See 
https://leanprover-community.github.io/mathlib4_docs/Init/Data/BitVec/Basic.html#BitVec.msb